src/System.Security.Claims/src/System/Security/Claims/ClaimsIdentity.cs

```diff
@@ -0,0 +1,970 @@
+        [SecurityCritical]
+        public virtual void RemoveClaim(Claim claim)
+        {
+            if (!TryRemoveClaim(claim))
+            {
+                throw new InvalidOperationException(
+                    string.Format(CultureInfo.InvariantCulture, 
+                                  SR.GetResourceString("InvalidOperation_ClaimCannotBeRemoved", "The Claim '{0}' was not able to be removed.  It is either not part of this Identity or it is a claim that is owned by the Principal that contains this Identity. For example, the Principal will own the claim when creating a GenericPrincipal with roles. The roles will be exposed through the Identity that is passed in the constructor, but not actually owned by the Identity.  Similar logic exists for a RolePrincipal."),
+                                  claim));
+            }
+        }
+
+        /// <summary>
+        /// Adds claims to intenal list. Calling Claim.Clone if Claim.Subject != this.
+        /// </summary>
+        /// <param name="claims">a <see cref="IEnumerable<Claim>"/> to add to </param>
+        /// <remarks>private only call from constructor, adds to internal list.</remarks>
+        [SecuritySafeCritical]
+        private void SafeAddClaims(IEnumerable<Claim> claims)
+        {
+            foreach (Claim claim in claims)
+            {
+                if (claim == null)
+                    continue;
+
+                if (object.ReferenceEquals(claim.Subject, this))
+                {
+                    _instanceClaims.Add(claim);
+                }
+                else
+                {
+                    _instanceClaims.Add(claim.Clone(this));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Adds claim to intenal list. Calling Claim.Clone if Claim.Subject != this.
+        /// </summary>
+        /// <remarks>private only call from constructor, adds to internal list.</remarks>
+        [SecuritySafeCritical]
+        private void SafeAddClaim(Claim claim)
+        {
+            if (claim == null)
+                return;
+
+            if (object.ReferenceEquals(claim.Subject, this))
+            {
+                _instanceClaims.Add(claim);
+            }
+            else
+            {
+                _instanceClaims.Add(claim.Clone(this));
+            }
+        }
+
+        /// <summary>
+        /// Retrieves a <see cref="IEnumerable{Claim}"/> where each claim is matched by <param name="match"/>.
+        /// </summary>
+        /// <param name="match">The function that performs the matching logic.</param>
+        /// <returns>A <see cref="IEnumerable{Claim}"/> of matched claims.</returns>
+        /// <exception cref="ArgumentNullException">if 'match' is null.</exception>
+        public virtual IEnumerable<Claim> FindAll(Predicate<Claim> match)
+        {
+            if (match == null)
+            {
+                throw new ArgumentNullException("match");
```

Comment:
This exception will be thrown out of the enumerator's MoveNext rather than out of the call to FindAll.  Is that intentional?  (Same comment on some of the other iterators.)